src/spikeinterface/core/analyzer_extension_core.py

```diff
@@ -248,6 +318,92 @@ def get_waveforms_one_unit(
+        kept_spikes : spikes, default None
+            If specified, the spikes associated with the kept waveforms
+
+        Returns
+        -------
+        some_waveforms: np.array
+            The waveforms (num_spikes, num_samples, num_sparse_channels)
+        spike_indices: np.array
+            Spike indices of the returned waveforms of shape (num_spikes, )
+        """
+
+        sorting = self.sorting_analyzer.sorting
+        if unit_ids is None:
+            unit_ids = sorting.unit_ids
+
+        if channel_ids is None:
+            channel_ids = self.sorting_analyzer.channel_ids
+
+        channel_indices = self.sorting_analyzer.channel_ids_to_indices(channel_ids)
+
+        # note : internally when sparse PCA are not aligned!! Exactly like waveforms.
+
+        sparsity = self.sorting_analyzer.sparsity
+
+        if kept_waveforms is not None:
+            waveforms = kept_waveforms
+        else:
+            waveforms = self.data["waveforms"]
+
+        num_samples = waveforms.shape[1]
+        dtype = waveforms.dtype
+
+        if kept_spikes is not None:
+            some_spikes = kept_spikes
+        else:
+            some_spikes = self.sorting_analyzer.get_extension("random_spikes").get_random_spikes()
+
+        unit_indices = sorting.ids_to_indices(unit_ids)
+        selected_inds = np.flatnonzero(np.isin(some_spikes["unit_index"], unit_indices))
+        spike_unit_indices = some_spikes["unit_index"][selected_inds]
+
+        if sparsity is None:
+            some_waveforms = waveforms[selected_inds, :, :][:, :, channel_indices]
+        else:
+            # need re-alignement
+            some_waveforms = np.zeros((selected_inds.size, num_samples, channel_indices.size), dtype=dtype)
+
+            for unit_id in unit_ids:
+                unit_index = sorting.id_to_index(unit_id)
+                sparse_waveforms = self.get_waveforms_one_unit(
+                    unit_id, kept_waveforms=kept_waveforms, kept_spikes=kept_spikes
+                )
+                local_chan_inds = sparsity.unit_id_to_channel_indices[unit_id]
+
+                # keep only requested channels
+                channel_mask = np.isin(local_chan_inds, channel_indices)
+                if sum(channel_mask) != len(channel_mask):
+                    sparse_waveforms = sparse_waveforms[:, :, channel_mask]
+
+                # inject in requested channels
+                spike_mask = np.flatnonzero(spike_unit_indices == unit_index)
+                channel_mask_channel_inds = np.isin(channel_indices, local_chan_inds)
+                if sum(channel_mask_channel_inds) < len(channel_mask_channel_inds):
+                    # inject in requested channels
+                    wfs = np.zeros((spike_mask.size, num_samples, channel_indices.size), dtype=dtype)
+                    wfs[:, :, channel_mask_channel_inds] = sparse_waveforms
+                else:
+                    wfs = sparse_waveforms
+                some_waveforms[spike_mask] = wfs
```

Comment:
This is should be outside in helper function.
Herte we have a weird design.

Reply:
This is exactly the same design that used to be there for the PC.